classes/migration/upgrade/v3_4_0/PreflightCheckMigration.inc.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/**
+ * @file classes/migration/upgrade/v3_4_0/PreflightCheckMigration.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class I6093_AddForeignKeys
+ * @brief Check for common problems early in the upgrade process.
+ */
+
+namespace PKP\migration\upgrade\v3_4_0;
+
+use Illuminate\Support\Facades\DB;
+
+use PKP\db\DAORegistry;
+
+class PreflightCheckMigration extends \PKP\migration\Migration
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        try {
+            // pkp/pkp-lib#6903 Prepare to add foreign key relationships
+            // Clean orphaned assoc_type/assoc_id data in announcement_types
+            $contextDao = \APP\core\Application::getContextDAO();
+            $orphanedIds = DB::table('announcement_types AS at')->leftJoin($contextDao->tableName . ' AS c', 'at.assoc_id', '=', 'c.' . $contextDao->primaryKeyColumn)->whereNull('c.' . $contextDao->primaryKeyColumn)->orWhere('at.assoc_type', '<>', \Application::get()->getContextAssocType())->distinct()->pluck('at.type_id');
+            foreach ($orphanedIds as $typeId) {
+                error_log("Removing orphaned announcement type ID ${typeId} with no matching context ID.");
```

Comment:
I'm not sure if it's useful to have these atomic logs with IDs. Then I think it's better to just [remove without listing](https://github.com/pkp/pkp-lib/pull/7279/files#diff-00cbf4c41a927a521d0edbb75ce310a7d68fab5c72f6a14d5328841478095c5eR45).

Reply:
At least until we're more confident about supporting end users through data correction, I'd like to err on the side of being more verbose rather than less. (But I did move these to `installer->log` rather than `error_log`.)